### DISCUSSION TITLE
Save presets with samplers in a directory (as option)

### DISCUSSION THREAD

**Discussion by Ampalmira** [publicsamples/Oi-Grandad#40] [Ideas]
Hi, I'm using the 1..1.4 (?) version of this fantastic pluging but I miss an essential option.

As Oi-Grandad needs samples to generate his fantastic sounds, I think that, the 'preset' term must contain not only the params but also the samples that need.

For this, I suggest (may be as an optional param in configuration) that the user can be able to save not only the 'preset' but also the samples used in this preset.

This kind of VST I use as an 'playground' where a get a sample from this, may be a fressound.org sample from donwload or another from a just donwloaded and unzipped directory located in a random folder that I will missed/deleted in a few days....

I suggest, as Waves CR-8 sampler does, to save the preset as always but create the option to save in the same directory the samples inside a directory named with the SAME that the preset.

In the next months I will be able to backup all my presets easyly because all the files will be in a 'know' place. No need to keep in mind where are the directories of the samples.

What do you think?

**Comment by Ampalmira** [comment]
... If you check your sample presets, some of them don't work because wav files are missing....

With the proposed suggestion these files will be inside the same folder as the presets inside their corresponding subfolders...

**Comment by RickTaylor79** [comment]
Thanks for the suggestion. It's a good idea, I'm not sure how I'd implement it just yet though. It could be a while before a feature like this gets added but it's now on the to-do-list.  

For the time being, it's probably best to keep all your samples in the modularsamples/AudioFiles folder. This way, should you want to backup, or share your presets with someone else, you can simply update the presets by opening them with a text editor and updating the paths as needed.

Here's a typical sample path entry in a preset:

` <Control type="ScriptAudioWaveform" id="AudioWaveform" value="0.0" data="/Users/YOURNAME/Library/Application Support/modularsamples/AudioFiles/Loops/Tappers-04.wav"`

We can make this more portable by replacing the path with the {AUDIO_FIILES} wildcard: 

` <Control type="ScriptAudioWaveform" id="AudioWaveform" value="0.0" data="{AUDIO_FILES}Loops/Tappers-04.wav"`

This will always point to the modularsamples/AudioFiles/ folder regardless of which system, or OS it is on.

So by using a text editor with the ability 'find and replace' across multiple files, or the Mac/Linux[ grep command](https://stackoverflow.com/questions/15402770/how-to-search-and-replace-using-grep), you can quickly replace all instances of "Users/YOURNAME/Library/Application Support/modularsamples/AudioFiles/" with "{AUDIO_FIILES}".

**Reply by Ampalmira** [reply]
hI Rick!

Thanks for your reply!

Yes, until this feature will come, the only way to keept a compack preset is placing all audio files into an know dir (if it is the only place better...).

Yes, I think that the use of {AUDIO_FIILES} avoid the usual headaches with the filesystems of different OS's....

I have NO idea how OI-grandad is created (I tried to diging into the code) but I think that the 'saving a preset' is a native or well known script made in HISE, right?

Sorry if I say something wrong or crazy but if 'saving a preset' is via HISE, may be you can experiment with the 'file' object of his scripting API.

https://docs.hise.audio/scripting/scripting-api/file/index.html

I see: 

- createDirectory (it is used when you create a user directory, so is well know from your side)
- copy (copies the file, as there are no extra params I assume that is a file located in an object saved in the 'current' directory....no way to tell 'where' you want to save the file...
- deleteFileOrDirectory : You need clear the directory of a preset if I rewritting with, may be, new samples. Better clear the directory and save the 'new' files....
- hasWriteAccess : Interesting for check if the dir is writeable or not, may be to avoid errors (I don't know if HISE show you a log or window telling for it).
- toString : Here, with the 'Filename' param passed you will isolate the name + extension of the file (may be you have this variable stored in a table when OI is running...)
- writeAudioFile : If OI are working with the audiofiles located in a buffer, this will be the ok for use it!

So for creating a presset:
1.- Store in a var the option if the 'save preset with sounds in a directory' is ON or OFF. If you think that this feature should be not mandatory and user can decide.
1.a - Read for this variable
2.- If it is OFF, saves preset as ever.
3.- If not, saves the preset as ever & copy the name of the preset as a str var
4.- Creates the dir with createDirectory
5.- Deletes the content of this directory (no matter if it is empty or not, to prevent garbage in this directory of unnused files) with deleteFileOrDirectory
6.- Creates a loop with the audio buffers or the object where the audio are stored and writes the content of the audio files in the located directory with writeAudioFile with the names of the original samples

I don't know if this help you or not. Sorry if I wrong...

**Reply by RickTaylor79** [reply]
Thanks for this! Yes, OG was built entirely with HISE. The preset management system used is the generic one built into HISE, so this would be quite a bit of work.  Now that you mention it, I recall doing some previous work trying to save audio buffers, it wasn't entirely successful but I didn't spend a great deal of time with it. There's no reason it shouldn't work with a bit of effort. 

It may be some time before I get to this as other stuff is piling up, feel free to play around with it if you have the inclination. You seem to know what you're talking about :)
